### DISCUSSION TITLE
Salesforce GiveWP integration.

### DISCUSSION THREAD

**Discussion by merguk** [impress-org/givewp#6687] [Developer Q&A] [answered]
We've followed these steps
https://givewp.com/documentation/add-ons/salesforce/sending-custom-data-to-salesforce/#scenario-2-adding-a-form-field-manager-field-to-a-custom-salesforce-field

Using the scenario 2 custom fields example, we've got this working

```
/**
 * @param  array  $args  - associate array of data to be sent to Salesforce. example: ["Field_Name" => "Value"]
 * @param  SalesforceOpportunityData  $data  - the original data object where you can access things like donationId. example: $data->donationId
 * @return array
 */
function addGiveSalesforceDonationArgs(
    $args,
    \GiveSalesforce\Salesforce\DataTransferObjects\SalesforceOpportunityData $data
) {
    // get the custom field from meta using the donationId from our $data DTO
    $title = give_get_meta(
        $data->donationId,
        'title',
        true
    );

    $company = give_get_meta(
        $data->donationId,
        'company',
        true
    );
    $street_address = give_get_meta(
        $data->donationId,
        'street_address',
        true
    );
    $suburb = give_get_meta(
        $data->donationId,
        'suburb',
        true
    );
    $postcode = give_get_meta(
        $data->donationId,
        'postcode',
        true
    );
    $state = give_get_meta(
        $data->donationId,
        'state',
        true
    );
    $country = give_get_meta(
        $data->donationId,
        'country',
        true
    );
    $phone_number = give_get_meta(
        $data->donationId,
        'phone_number',
        true
    );
    $where_did_you_hear_about_safe_steps = give_get_meta(
        $data->donationId,
        'where_did_you_hear_about_safe_steps',
        true
    );
    $heard_about___other = give_get_meta(
        $data->donationId,
        'heard_about___other',
        true
    );
    $why_are_you_donating_today = give_get_meta(
        $data->donationId,
        'why_are_you_donating_today',
        true
    );


    // assign the Salesforce custom field name to the GiveWP value

// These assignments don't work - assume it's because we need a separate arg and function for Contacts object. 
    $args['Salutation'] = $title;
    $args['OtherStreet'] = $street_address;
    $args['OtherCity'] = $suburb;
    $args['OtherState'] = $state;
    $args['OtherPostalCode'] = $postcode;
    $args['OtherCountry'] = $country;
    $args['Description'] = $company;
    $args['HomePhone'] = $phone_number;

// these assignments work, as they are on the Opportunity object
    $args['Where_did_you_hear_about_Safe_Steps__c'] = $where_did_you_hear_about_safe_steps;
    $args['Comments__c'] = $heard_about___other;
    $args['Why_are_you_donating_today__c'] = $why_are_you_donating_today;

    // return the modified array
    return $args;
}

add_filter('give_salesforce_donation_args', 'addGiveSalesforceDonationArgs', 10, 2);
```

Can we extend the logic you have documented, to Contacts? I believe the key is knowing the filter API.

You have documented:

**Filter name:**
`give_salesforce_donation_args`
**Filter API**
`apply_filters('give_salesforce_donation_args', [], \GiveSalesforce\Salesforce\DataTransferObjects\SalesforceOpportunityData $salesforceOpportunityData);`

This comment

> type-hinting SalesforceOpportunityData will autocomplete values in most code editors when accessing properties from the $data object

is key, and will probably help me find the Contact equivalent, but I don't have admin access to this site, so I am hoping someone else can tell me here.

So I am thinking additional filter names and filter API for Contact.  Here's a guess, could anyone verify?

**Filter name:**
`give_salesforce_CONTACT_args`
**Filter API**
`apply_filters('give_salesforce_CONTACT_args', [], \GiveSalesforce\Salesforce\DataTransferObjects\SalesforceCONTACTData $salesforceCONTACTData);`

Would the add filter be the same format too? Including the 10,2 parameters?

example documented for Opportunity:
`add_filter('give_salesforce_donation_args', 'addGiveSalesforceDonationArgs', 10, 2);`

My guess:
`add_filter('give_salesforce_CONTACT_args', 'addGiveSalesforceCONTACTArgs', 10, 2);`

**Comment by jonwaldstein** [comment, marked as answer]
Hi @merguk

You are correct, these filter args will only work on the opportunity record.  We currently do not have a filter for contact args.  I would encourage you to submit a feature request [here](https://givewp.canny.io/feature-requests) so we can track interest and hopefully get this added.
